File: tks/ApplicationCore/ApplicationServices/src/test/kotlin/services/RentServiceTest.java
package services;

import exceptions.*;
import drivenports.RentManagePort;
import drivenports.RentQueryPort;
import ports.secondary.ResourceSearchPort;
import ports.secondary.UserSearchPort;
import model.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Date;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class RentServiceTest {

    RentService rentService;

    Resource sampleResource;
    User sampleUser;
    Rent sampleRent;
    String sampleEmail;
    String sampleEmail2;
    UUID sampleResId;
    UUID sampleRentId;

    @Mock
    ResourceSearchPort resourceSearchPort;
    @Mock
    UserSearchPort userSearchPort;
    @Mock
    RentQueryPort rentQueryPort;
    @Mock
    RentManagePort rentManagePort;

    @BeforeEach
    public void init(){
        sampleEmail = "mszewc@edu.pl";
        sampleEmail2 = "mzab@edu.pl";
        sampleResId = UUID.randomUUID();
        sampleRentId = UUID.randomUUID();
        sampleUser = new User(sampleEmail, UserRole.CLIENT, "####", true);
        sampleResource = new Book(sampleResId, "Diuna", "Frank Herbert");
        sampleRent = new Rent(sampleRentId, new Date(), null, sampleEmail2, sampleResId);
        rentService = new RentService(rentManagePort, rentQueryPort, userSearchPort, resourceSearchPort);
    }

    @Test
    public void GivenInvalidUserId_Rent_ShouldFail(){
        when(userSearchPort.findByEmail(eq(sampleUser.getEmail()))).thenReturn(null);
        assertThrows(UserNotFoundException.class, () -> rentService.rent(sampleUser.getEmail(), sampleResource.getId()));
        verify(rentManagePort, never()).save(any());
    }

    @Test
    public void GivenInvalidResourceId_Rent_ShouldFail(){
        when(userSearchPort.findByEmail(eq(sampleEmail))).thenReturn(sampleUser);
        when(resourceSearchPort.findById(eq(sampleResId))).thenReturn(null);
        assertThrows(ResourceNotFoundException.class, () -> rentService.rent(sampleUser.getEmail(), sampleResource.getId()));
        verify(rentManagePort, never()).save(any());
    }

    @Test
    public void GivenNonActiveUserId_Rent_ShouldFail(){
        sampleUser.setActive(false);
        when(userSearchPort.findByEmail(eq(sampleEmail))).thenReturn(sampleUser);
        when(resourceSearchPort.findById(eq(sampleResId))).thenReturn(sampleResource);
        assertThrows(UserNotActiveException.class, () -> rentService.rent(sampleUser.getEmail(), sampleResource.getId()));
        verify(rentManagePort, never()).save(any());
    }

    @Test
    public void GivenAlreadyRentResId_Rent_ShouldFail(){
        when(userSearchPort.findByEmail(eq(sampleEmail))).thenReturn(sampleUser);
        when(resourceSearchPort.findById(eq(sampleResId))).thenReturn(sampleResource);
        when(rentQueryPort.findActiveByResourceId(sampleResId)).thenReturn(sampleRent);
        assertThrows(ResourceAlreadyRentException.class, () -> rentService.rent(sampleUser.getEmail(), sampleResource.getId()));
        verify(rentManagePort, never()).save(any());
    }

    @Test
    public void GivenInvalidUserId_Return_ShouldFail(){
        when(userSearchPort.findByEmail(eq(sampleEmail))).thenReturn(null);
        assertThrows(UserNotFoundException.class, () -> rentService.returnResource(sampleEmail, sampleResId));
    }

    @Test
    public void GivenInvalidResourceId_Return_ShouldFail(){
        when(userSearchPort.findByEmail(eq(sampleEmail))).thenReturn(sampleUser);
        when(resourceSearchPort.findById(eq(sampleResId))).thenReturn(null);
        assertThrows(ResourceNotFoundException.class, () -> rentService.returnResource(sampleEmail, sampleResId));
    }

    @Test
    public void GivenNotRentResourceId_Return_ShouldFail(){
        when(userSearchPort.findByEmail(eq(sampleEmail))).thenReturn(sampleUser);
        when(resourceSearchPort.findById(eq(sampleResId))).thenReturn(sampleResource);
        when(rentQueryPort.findActiveByResourceId(sampleResId)).thenReturn(null);
        assertThrows(ResourceNotRentException.class, () -> rentService.returnResource(sampleEmail, sampleResId));
    }

    @Test
    public void GivenResourceRentByOtherUser_Return_ShouldFail(){
        when(userSearchPort.findByEmail(eq(sampleEmail))).thenReturn(sampleUser);
        when(resourceSearchPort.findById(eq(sampleResId))).thenReturn(sampleResource);
        sampleRent = new Rent(sampleRentId, new Date(), null, sampleEmail2, sampleResId);
        when(rentQueryPort.findActiveByResourceId(sampleResId)).thenReturn(sampleRent);
        assertThrows(InvalidUserException.class, () -> rentService.returnResource(sampleEmail, sampleResId));
    }

}

File: tks/ApplicationCore/ApplicationServices/src/test/kotlin/application/services/RentServiceTest.kt
//package application.services
//
//import application.helpers.AccessionNumberHelper
//import domain.exceptions.*
//import domain.model.context.library.Book
//import domain.model.context.rents.Rent
//import domain.model.context.users.User
//import domain.model.UserRole
//import domain.model.context.library.Resource
//import domain.model.values.AccessionNumber
//import domain.model.values.Email
//import io.mockk.every
//import io.mockk.impl.annotations.RelaxedMockK
//import io.mockk.junit5.MockKExtension
//import io.mockk.verify
//import org.junit.jupiter.api.Assertions
//import org.junit.jupiter.api.BeforeEach
//import org.junit.jupiter.api.Test
//import org.junit.jupiter.api.extension.ExtendWith
//import ports.secondary.RentPersistencePort
//import ports.secondary.RentSearchPort
//import ports.secondary.ResourceSearchPort
//import ports.secondary.UserSearchPort
//import java.util.*
//
//@ExtendWith(MockKExtension::class)
//internal class RentServiceTest {
//    private lateinit var rentService: RentService
//    private lateinit var sampleResource: Resource
//    private lateinit var sampleUser: User
//    private lateinit var sampleRent: Rent
//    private val sampleEmail: Email = Email("mszewc@edu.pl")
//    private val sampleEmail2: Email = Email("mzab@edu.pl")
//    private val sampleResId: AccessionNumber = AccessionNumberHelper.generate()
//    private val sampleRentId: UUID = UUID.randomUUID()
//
//    @RelaxedMockK
//    lateinit var resourceSearchPort: ResourceSearchPort
//
//    @RelaxedMockK
//    lateinit var userSearchPort: UserSearchPort
//
//    @RelaxedMockK
//    lateinit var rentSearchPort: RentSearchPort
//
//    @RelaxedMockK
//    lateinit var rentPersistencePort: RentPersistencePort
//
//    @BeforeEach
//    fun init() {
//        sampleUser = User(sampleEmail, UserRole.CLIENT, "####", true)
//        sampleResource = Book(sampleResId, "Diuna", "Frank Herbert")
//        sampleRent = Rent(sampleRentId, Date(), null, sampleEmail2, sampleResId)
//        rentService = RentService(rentPersistencePort, rentSearchPort, userSearchPort, resourceSearchPort)
//    }
//
//    @Test
//    fun `Given invalid user id then rent should fail`() {
//        every { userSearchPort.findByEmail(sampleUser.email) } returns null
//        Assertions.assertThrows(UserNotFoundException::class.java) {
//            rentService.rent(
//                sampleUser.email, sampleResource.accessionNumber!!
//            )
//        }
//        verify(exactly = 0) { rentPersistencePort.save(any()) }
//    }
//
//    @Test
//    fun `Given invalid resource id then rent should fail`() {
//        every { userSearchPort.findByEmail(sampleEmail) } returns sampleUser
//        every { resourceSearchPort.findByAccessionNumber(sampleResId) } returns null
//        Assertions.assertThrows(ResourceNotFoundException::class.java) {
//            rentService.rent(
//                sampleUser.email, sampleResource.accessionNumber!!
//            )
//        }
//        verify(exactly = 0) { rentPersistencePort.save(any()) }
//    }
//
//    @Test
//    fun `Given inactive user email then rent should fail`() {
//        sampleUser.active = false
//        every { userSearchPort.findByEmail(sampleEmail) } returns sampleUser
//        every { resourceSearchPort.findByAccessionNumber(sampleResId) } returns sampleResource
//        Assertions.assertThrows(UserNotActiveException::class.java) {
//            rentService.rent(
//                sampleUser.email, sampleResource.accessionNumber!!
//            )
//        }
//        verify(exactly = 0) { rentPersistencePort.save(any()) }
//    }
//
//    @Test
//    fun `Given already rent resource id then rent should fail`() {
//        every { userSearchPort.findByEmail(sampleEmail) } returns sampleUser
//        every { resourceSearchPort.findByAccessionNumber(sampleResId) } returns sampleResource
//        every { rentSearchPort.findActiveByResourceId(sampleResId) } returns sampleRent
//        Assertions.assertThrows(ResourceAlreadyRentException::class.java) {
//            rentService.rent(
//                sampleUser.email, sampleResource.accessionNumber!!
//            )
//        }
//        verify(exactly = 0) { rentPersistencePort.save(any()) }
//    }
//
//    @Test
//    fun `Given invalid user id then returnResource should fail`() {
//        every { userSearchPort.findByEmail(sampleEmail) } returns null
//        Assertions.assertThrows(UserNotFoundException::class.java) {
//            rentService.returnResource(
//                sampleEmail,
//                sampleResId
//            )
//        }
//    }
//
//    @Test
//    fun `Given invalid resource id then returnResource should fail`() {
//        every { userSearchPort.findByEmail(sampleEmail) } returns sampleUser
//        every { resourceSearchPort.findByAccessionNumber(sampleResId) } returns null
//        Assertions.assertThrows(ResourceNotFoundException::class.java) {
//            rentService.returnResource(
//                sampleEmail,
//                sampleResId
//            )
//        }
//    }
//
//    @Test
//    fun `Given not rent resource id returnResource should fail`() {
//        every { userSearchPort.findByEmail(sampleEmail) } returns sampleUser
//        every { resourceSearchPort.findByAccessionNumber(sampleResId) } returns sampleResource
//        every { rentSearchPort.findActiveByResourceId(sampleResId) } returns null
//        Assertions.assertThrows(ResourceNotRentException::class.java) {
//            rentService.returnResource(
//                sampleEmail,
//                sampleResId
//            )
//        }
//    }
//
//    @Test
//    fun `Given resource rent by other user then returnResource should fail`() {
//        every { userSearchPort.findByEmail((sampleEmail)) } returns sampleUser
//        every { resourceSearchPort.findByAccessionNumber((sampleResId)) } returns sampleResource
//        sampleRent = Rent(sampleRentId, Date(), null, sampleEmail2, sampleResId)
//        every { rentSearchPort.findActiveByResourceId(sampleResId) } returns sampleRent
//        Assertions.assertThrows(InvalidUserException::class.java) {
//            rentService.returnResource(
//                sampleEmail,
//                sampleResId
//            )
//        }
//    }
//
//    @Test
//    fun `Given valid id then getDetails should return rent`() {
//        val rent = Rent(sampleRentId, Date(), null, sampleEmail, sampleResId)
//        every { (rentSearchPort.getById(sampleRentId)) }returns rent
//        val result = rentService.getDetails(sampleRentId)
//        Assertions.assertEquals(result, rent)
//    }
//
//    @Test
//    fun `Given invalid id then getDetails should fail`() {
//        every { rentSearchPort.getById(sampleRentId) }returns null
//        Assertions.assertThrows(RentNotFoundException::class.java) { rentService.getDetails(sampleRentId) }
//    }
//}